src/can.rs

```diff
@@ -0,0 +1,1060 @@
+            Id::Standard(id) if id <= 0x7FF => true,
+            Id::Extended(id) if id <= 0x1FFF_FFFF => true,
+            _ => false,
+        }
+    }
+}
+
+/// Identifier of a CAN message.
+///
+/// Can be either a standard identifier (11bit, Range: 0..0x3FF) or a
+/// extendended identifier (29bit , Range: 0..0x1FFFFFFF).
+///
+/// The `Ord` trait can be used to determine the frame’s priority this ID
+/// belongs to.
+/// Lower identifier values have a higher priority. Additionally standard frames
+/// have a higher priority than extended frames and data frames have a higher
+/// priority than remote frames.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+struct IdReg(u32);
+
+impl IdReg {
+    const STANDARD_SHIFT: u32 = 21;
+    const STANDARD_MASK: u32 = 0x7FF << Self::STANDARD_SHIFT;
+
+    const EXTENDED_SHIFT: u32 = 3;
+    const EXTENDED_MASK: u32 = 0x1FFF_FFFF << Self::EXTENDED_SHIFT;
+
+    const IDE_MASK: u32 = 0x0000_0004;
+
+    const RTR_MASK: u32 = 0x0000_0002;
+
+    /// Creates a new standard identifier (11bit, Range: 0..0x7FF)
+    ///
+    /// Panics for IDs outside the allowed range.
+    pub fn new_standard(id: u32) -> Self {
+        assert!(id < 0x7FF);
+        Self(id << Self::STANDARD_SHIFT)
+    }
+
+    /// Creates a new extendended identifier (29bit , Range: 0..0x1FFFFFFF).
+    ///
+    /// Panics for IDs outside the allowed range.
+    fn new_extended(id: u32) -> IdReg {
+        assert!(id < 0x1FFF_FFFF);
+        Self(id << Self::EXTENDED_SHIFT | Self::IDE_MASK)
+    }
+
+    fn from_register(reg: u32) -> IdReg {
+        Self(reg & 0xFFFF_FFFE)
+    }
+
+    /// Sets the remote transmission (RTR) flag. This marks the identifier as
+    /// being part of a remote frame.
+    fn with_rtr(self, rtr: bool) -> IdReg {
+        if rtr {
+            Self(self.0 | Self::RTR_MASK)
+        } else {
+            Self(self.0 & !Self::RTR_MASK)
+        }
+    }
+
+    /// Returns the identifier.
+    fn to_id(self) -> Id {
+        if self.is_extended() {
+            Id::Extended(self.0 >> Self::EXTENDED_SHIFT)
+        } else {
+            Id::Standard(self.0 >> Self::STANDARD_SHIFT)
+        }
+    }
+
+    /// Returns `true` if the identifier is an extended identifier.
+    fn is_extended(self) -> bool {
+        self.0 & Self::IDE_MASK != 0
+    }
+
+    /// Returns `true` if the identifier is a standard identifier.
+    fn is_standard(self) -> bool {
+        !self.is_extended()
+    }
+
+    /// Returns `true` if the identifer is part of a remote frame (RTR bit set).
+    fn rtr(self) -> bool {
+        self.0 & Self::RTR_MASK != 0
+    }
+}
+
+impl Ord for IdReg {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match (self.is_standard(), other.is_standard()) {
+            (true, false) => Ordering::Less,
+            (false, true) => Ordering::Greater,
+            // Ordering of the data/remote frames implicitly gives by the bit layout.
+            _ => self.0.cmp(&other.0),
+        }
```

Comment:
I don't think this computation is correct. Standard frames only have priority over extended frames when the most significant 11 bits of the IDs match.

Reply:
Thank you that is a great catch! I’m gonna fix this and include the data in the `PartialOrd` so that it matches `PartialEq`.